feat: implement webscoket and webhook handlers

Extheoisah opened the pull request:
This pr adds a WebSocket server in addition to the web server. It also adds a webhook endpoint for receiving events from the git service and the test runners.
The git service webhook implementation is complete and can be tested. However, the test runners webhook implementation is still  WIP. When the test runner work is complete in the test runners repository, we will finish the implementation on the server side.
However, that is non-blocking and this PR can merged regardless.

Extheoisah committed:
feat: initialize webscoket and webhook handlers

Extheoisah committed:
chore: add wscat for testing websocket

Extheoisah committed:
feat: implment webhook to websocket data stream

Extheoisah committed:
feat(docs): add websocket and webhook documentation

daviroo commented on src/app/routes/mod.rs:
```diff
@@ -5,6 +5,7 @@ pub mod health;
 pub mod signin;
 pub mod signup;
 pub mod users;
+pub mod webhook;
```
Do we need to initialise the webhook in the `init` function in this file?

Extheoisah replied:
I don't want the webhook route to live behind the "/api" scope that was why I didn't include it here

daviroo commented on src/app/websockets/handler.rs:
```diff
@@ -0,0 +1,120 @@
+            log::info!("Binary message received: {:?}", bin);
+            let message = Message::Binary(bin);
+            manager_handle
+                .send_message_to_session(session_token, message, conn_id)
```
Because of the way the `websocket_message_to_bytestring` [func](https://github.com/Extheoisah/hxckr-core/blob/ae3f14f680e39260139b543f790c0542a7f01b78/src/shared/utils.rs#L21) has been implemented if we pass a `Binary` type message to the `send_message_to_session` function it will be converted to an empty string.  I think we either need to handle binary type messages in `send_message_to_session` or perhaps we want to exclude them entirely if we know we will never use them on the client.  What do you think?

daviroo replied:
One suggestion for a way to handle it in `send_message_to_session`:

```rust
pub async fn send_message_to_session(
    &self,
    session_token: &SessionToken,
    message: Message,
    sender_conn_id: ConnId,
) -> io::Result<()> {
    let sessions = self.sessions.read().await;
    let connections = self.connections.read().await;

    if let Some(conn_ids) = sessions.get(session_token) {
        for &conn_id in conn_ids {
            log::info!(
                "sender_conn_id: {:?}, conn_id: {:?}",
                sender_conn_id,
                conn_id
            );
            if let Some(conn) = connections.get(&conn_id) {
                let mut session = conn.session.clone();
                match message {
                    Message::Text(ref text) => {
                        if let Err(e) = session.text(text).await {
                            log::error!("Failed to send text message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    Message::Binary(ref bytes) => {
                        if let Err(e) = session.binary(bytes.clone()).await {
                            log::error!("Failed to send binary message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    Message::Ping(ref bytes) => {
                        if let Err(e) = session.ping(bytes.clone()).await {
                            log::error!("Failed to send ping message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    Message::Pong(ref bytes) => {
                        if let Err(e) = session.pong(bytes.clone()).await {
                            log::error!("Failed to send pong message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    Message::Close(reason) => {
                        if let Err(e) = session.close(reason).await {
                            log::error!("Failed to send close message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    // Handle other message types if necessary
                }
            }
        }
    }
    Ok(())
}
```

Extheoisah replied:
i don't think we'll need to send or receive binary messages as we'll be sending just JSON text to the client which was why I only handled the text case. 
But I have refactored it to account for binary messages. Its not robust because it doesn't account for complex binary decoding. Given we don't have a use case for it now, I think it's not a deal breaker.

daviroo commented on src/app/websockets/handler.rs:
```diff
@@ -0,0 +1,120 @@
+                .await?;
+        }
+        Message::Close(reason) => {
+            log::info!("Close message received: {:?}", reason);
+            return Err(Error::from(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "WebSocket closed",
+            )));
```
```suggestion
            log::info!("Close message received: {:?}", reason);
            manager_handle.disconnect(conn_id).await?;
            session.close(reason).await.map_err(|e| Error::from(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
            return Ok(());  // Return Ok to gracefully exit the WebSocket loop
```

daviroo commented on src/service/repository/repo.rs:
```diff
@@ -78,7 +79,23 @@ impl Repository {
+                Ok(repo)
+            }
+            (None, None, None) => Err(anyhow::anyhow!("No input provided")),
             _ => Err(anyhow::anyhow!("Invalid input")),
```
```suggestion
            (Some(_), Some(_), None) | (Some(_), None, Some(_)) | (None, Some(_), Some(_)) => {
                Err(anyhow::anyhow!("Only one of id, user_id, or repo_url should be provided"))
            }
            (Some(_), Some(_), Some(_)) => {
                Err(anyhow::anyhow!("Cannot provide id, user_id, and repo_url simultaneously"))
            }
```

This change improves input validation by explicitly handling all possible input combinations and providing more descriptive error messages. It enhances the function's robustness, makes its behavior more predictable, and facilitates easier debugging by clearly communicating which input combinations are invalid and why.

Extheoisah replied:
Nice. Added to last commit https://github.com/Extheoisah/hxckr-core/pull/17/commits/42d1b9f79e931dea9e862e025757782ae8906f32

daviroo reviewed (changes requested):
Some small improvements/suggestions.  The main thing I think we need to think about is how we are handling the different message types in `send_message_to_session` & `websocket_message_to_bytestring`.

Extheoisah committed:
feat: address review comments

1. account for binary message type
2. remove unused and redundant broadcast_to_all method
3. gracefully close websocket connection if a Close message is recieved
